src/code/sys_math3d.c

```diff
@@ -1,189 +1,2456 @@
+        intersect->point.x = 0.0f;
+    } else if ((dirY >= dirX) && (dirY >= dirZ)) {
+        Math3D_FindPointOnPlaneIntersect(planeAC, planeAA, planeBC, planeBA, intersect->dir.y, planeADist, planeBDist,
+                                         &intersect->point.z, &intersect->point.x);
+        intersect->point.y = 0.0f;
+    } else {
+        Math3D_FindPointOnPlaneIntersect(planeAA, planeAB, planeBA, planeBB, intersect->dir.z, planeADist, planeBDist,
+                                         &intersect->point.x, &intersect->point.y);
+        intersect->point.z = 0.0f;
+    }
+    return true;
+}
+
+/**
+ * Gets the closest point on the line formed from the intersection of of the planes defined from
+ * `planeAA`x + `planeAB`y + `planeAC`z + `planeADist` = 0 and
+ * `planeBA`x + `planeBB`y + `planeBC`z + `planeBDist` = 0
+ * the point on the intersection line closest to `point` is placed in `closestPoint`
+ * returns false if the planes are parallel.
+ */
+s32 Math3D_PlaneVsPlaneVsLineClosestPoint(f32 planeAA, f32 planeAB, f32 planeAC, f32 planeADist, f32 planeBA,
+                                          f32 planeBB, f32 planeBC, f32 planeBDist, Vec3f* point, Vec3f* closestPoint) {
+    static InfiniteLine sPlaneIntersect;
+
+    if (!Math3D_PlaneVsPlaneNewLine(planeAA, planeAB, planeAC, planeADist, planeBA, planeBB, planeBC, planeBDist,
+                                    &sPlaneIntersect)) {
+        return false;
+    }
+    Math3D_LineClosestToPoint(&sPlaneIntersect, point, closestPoint);
+    return true;
+}
+
+/**
+ * Calculates the point on the line from starting point `v0`, in the direction `dir` scaled by `scale`. Result is placed
+ * in `ret`
+ */
+void Math3D_PointOnDirectedLine(Vec3f* v0, Vec3f* dir, f32 scale, Vec3f* ret) {
+    ret->x = (dir->x * scale) + v0->x;
+    ret->y = (dir->y * scale) + v0->y;
+    ret->z = (dir->z * scale) + v0->z;
+}
+
+/**
+ * Splits the line segment from end points `v0` and `v1`, and splits that segment
+ * by `ratio` of `v0`:`v1`, places the resulting point on the line in `ret`
+ */
+void Math3D_LineSplitRatio(Vec3f* v0, Vec3f* v1, f32 ratio, Vec3f* ret) {
+    Vec3f diff;
+
+    Math_Vec3f_Diff(v1, v0, &diff);
+    Math3D_PointOnDirectedLine(v0, &diff, ratio, ret);
+}
+
+/**
+ * Calculates the cosine between vectors `a` and `b`
+ */
+f32 Math3D_Cos(Vec3f* a, Vec3f* b) {
+    f32 ret;
+
+    Math3D_CosOut(a, b, &ret);
+    return ret;
+}
+
+/**
+ * Calculates the cosine between vectors `a` and `b` and places the result in `ret`
+ * returns true if the cosine cannot be calculated because the product of the magnitudes is zero
+ */
+s32 Math3D_CosOut(Vec3f* a, Vec3f* b, f32* dst) {
+    f32 magProduct = Math3D_Vec3fMagnitude(a) * Math3D_Vec3fMagnitude(b);
+
+    if (IS_ZERO(magProduct)) {
+        *dst = 0.0f;
+        return true;
+    }
+    *dst = ((a->x * b->x) + (a->y * b->y) + (a->z * b->z)) / magProduct;
+    return false;
```

Comment:
Err, puts it in "dst"? I don't see a "ret"

Reply:
Oh funny, yeah that mistake is also in OoT